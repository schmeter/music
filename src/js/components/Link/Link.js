import React from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';


const isExternal = (path) => /^https?:\/\/.+/.test(path);

class Link extends React.Component {
    render() {
        const { className, to, onClick, children } = this.props;
        const props = {
            className,
            onClick
        };
        const content = children || to;
        return isExternal(to) ? (
            <a
                target="_blank"
                href={to}
                {...props}
            >
                {content}
            </a>
        ) : (
            <NavLink
                exact
                activeClassName="active"
                to={to}
                {...props}
            >
                {content}
            </NavLink>
        );
    }
}

Link.propTypes = {
    to: PropTypes.string.isRequired,
    children: PropTypes.node,
    onClick: PropTypes.func,
    className: PropTypes.string
};

export default Link;
