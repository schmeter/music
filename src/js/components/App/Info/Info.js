import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import Image from '../../Image';
import Link from '../../Link';
import i18n from '../../../services/i18n';
import { getUrl } from '../../../services/navigation';
import { scrollTop } from '../../../util/screen';

const infos = [
    { id: 'title' },
    { id: 'album', link: 'audio:artistId:albumId' },
    { id: 'artist', link: 'audio:artistId' },
];
const scrollClass = 'layer-content-info';

const Info = ({
    activeTrack,
    closeLayers,
}) => {
    useEffect(() => {
        scrollTop(`.${scrollClass}`);
    }, [
        activeTrack,
    ]);

    return !activeTrack ? null : (
        <div
            className={classNames(
                'layer-content',
                scrollClass,
            )}
        >
            {activeTrack.lyrics && (
                <div
                    className="lyrics"
                    dangerouslySetInnerHTML={{ __html: activeTrack.lyrics }}
                />
            )}
            <ul className="table">
                {infos.map(info => activeTrack.tag[info.id] && (
                    <li
                        key={info.id}
                        className="table-row"
                    >
                        <span className="table-cell text-left bold">
                            {i18n(`layer_info_${info.id}`)}:
                        </span>
                        <span className="table-cell text-left">
                            {info.link ? (
                                <Link
                                    to={getUrl(info.link, {
                                        artistId: activeTrack.artist.id,
                                        albumId: activeTrack.album.id,
                                    })}
                                    onClick={closeLayers}
                                >
                                    {activeTrack.tag[info.id]}
                                </Link>
                            ) : activeTrack.tag[info.id]}
                        </span>
                    </li>
                ))}
            </ul>
            <Link
                className="cover"
                to={getUrl(infos.find(info => info.id === 'album').link, {
                    artistId: activeTrack.artist.id,
                    albumId: activeTrack.album.id,
                })}
                onClick={closeLayers}
            >
                <Image
                    src={activeTrack.album.imgPath}
                    alt={activeTrack.album.title}
                />
            </Link>
            <audio
                className="spacer"
                controls
            />
        </div>
    );
};

Info.propTypes = {
    activeTrack: PropTypes.object,
    closeLayers: PropTypes.func.isRequired,
};

export default Info;
