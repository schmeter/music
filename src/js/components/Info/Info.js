import React, { useEffect } from 'react';
import PropTypes from 'prop-types';

import Image from '../Image';
import Link from '../Link';
import i18n from '../../helpers/i18n';
import { getRouteWithParams } from '../../helpers/navigation';
import { scrollTop } from '../../helpers/screen';

const Info = ({
  activeTrack,
  closeLayers,
}) => {
  useEffect(() => {
    scrollTop('.layer-info .layer-content');
  }, [
    activeTrack,
  ]);

  if (!activeTrack) {
    return null;
  }

  const { artist, album, lyrics, tag } = activeTrack;
  const infos = ['title', 'album', 'artist'];
  const links = {
    album: getRouteWithParams('audio', {
      artistId: artist.id,
      albumId: album.id,
    }),
  };

  return (
    <>
      {lyrics && (
        <div
          className="lyrics"
          dangerouslySetInnerHTML={{ __html: lyrics }}
        />
      )}
      <ul className="table">
        {infos.map(info => tag[info] && (
          <li
            key={info}
            className="table-row"
          >
            <span className="table-cell text-left bold">
              {i18n(`layer_info_${info}`)}:
            </span>
            <span className="table-cell text-left">
              {links[info] ? (
                <Link
                  to={links[info]}
                  onClick={closeLayers}
                >
                  {tag[info]}
                </Link>
              ) : tag[info]}
            </span>
          </li>
        ))}
      </ul>
      <Link
        className="cover"
        to={links.album}
        onClick={closeLayers}
      >
        <Image
          src={album.imgPath}
          alt={album.title}
        />
      </Link>
    </>
  );
};

Info.propTypes = {
  activeTrack: PropTypes.object,
  closeLayers: PropTypes.func.isRequired,
};

export default Info;
