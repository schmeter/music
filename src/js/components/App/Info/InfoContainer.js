import { connect } from 'react-redux';

import Info from './Info';
import { closeLayersAction } from '../../../state/actions/layer';

export const mapDispatchToProps = (dispatch) => ({
    closeLayers: () => dispatch(closeLayersAction())
});

export const mapStateToProps = (state) => ({
    activeTrack: state.audio.activeTrack
});

export default connect(mapStateToProps, mapDispatchToProps)(Info);
