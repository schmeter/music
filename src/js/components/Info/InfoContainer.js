import { connect } from 'react-redux';

import Info from './Info';
import { closeLayersAction } from '../../state/actions/layer';
import { getActiveTrack } from '../../state/selectors/audio';

export const mapDispatchToProps = dispatch => ({
  closeLayers: () => dispatch(closeLayersAction()),
});

export const mapStateToProps = state => ({
  activeTrack: getActiveTrack(state),
});

export default connect(mapStateToProps, mapDispatchToProps)(Info);
