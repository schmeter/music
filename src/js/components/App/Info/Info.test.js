import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Info from './Info';
import Link from '../../Link';
// import { scrollTop } from '../../../util/screen';

jest.mock('../../../util/screen', () => ({
    scrollTop: jest.fn(),
}));

describe('Info', () => {
    const mockedEvent = {
        preventDefault: jest.fn(),
    };

    it('renders correctly', () => {
        const activeTrack = {
            path: 'test',
            tag: {
                title: 'test',
            },
            artist: {
                id: 'test',
            },
            album: {
                id: 'test',
                title: 'test',
                imgPath: 'test',
            },
        };
        const lyrics = 'test';
        const props = {
            closeLayers: jest.fn(),
        };
        const component = shallow(<Info {...props} />);

        expect(toJson(component)).toMatchSnapshot();

        component.setProps({
            ...props,
            activeTrack,
        });
        component.update();

        // TODO: activate when enzyme supports hooks
        // expect(scrollTop).toHaveBeenCalled();

        component.setProps({
            ...props,
            activeTrack: {
                ...activeTrack,
                tag: {
                    title: 'test',
                    artist: 'test',
                    album: 'test',
                },
                lyrics,
            },
        });
        component.update();
        component.find('.cover').find(Link).simulate('click', mockedEvent);

        expect(toJson(component)).toMatchSnapshot();
        expect(props.closeLayers).toHaveBeenCalled();
    });
});
