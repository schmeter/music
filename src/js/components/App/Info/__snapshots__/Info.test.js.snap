// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Info renders correctly 1`] = `""`;

exports[`Info renders correctly 2`] = `
<div
  className="layer-content layer-content-info"
>
  <div
    className="lyrics"
    dangerouslySetInnerHTML={
      Object {
        "__html": "test",
      }
    }
  />
  <ul
    className="table"
  >
    <li
      className="table-row"
      key="title"
    >
      <span
        className="table-cell text-left bold"
      >
        Titel
        :
      </span>
      <span
        className="table-cell text-left"
      >
        test
      </span>
    </li>
    <li
      className="table-row"
      key="album"
    >
      <span
        className="table-cell text-left bold"
      >
        Album
        :
      </span>
      <span
        className="table-cell text-left"
      >
        <Link
          onClick={
            [MockFunction] {
              "calls": Array [
                Array [
                  Object {
                    "preventDefault": [MockFunction],
                  },
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            }
          }
          to="/audio/test/test"
        >
          test
        </Link>
      </span>
    </li>
    <li
      className="table-row"
      key="artist"
    >
      <span
        className="table-cell text-left bold"
      >
        KÃ¼nstler
        :
      </span>
      <span
        className="table-cell text-left"
      >
        <Link
          onClick={
            [MockFunction] {
              "calls": Array [
                Array [
                  Object {
                    "preventDefault": [MockFunction],
                  },
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            }
          }
          to="/audio/test"
        >
          test
        </Link>
      </span>
    </li>
  </ul>
  <Link
    className="cover"
    onClick={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "preventDefault": [MockFunction],
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    to="/audio/test/test"
  >
    <Image
      alt="test"
      src="test"
    />
  </Link>
  <audio
    className="spacer"
    controls={true}
  />
</div>
`;
